circuits/core/src/bhp/hash.rs

```diff
@@ -0,0 +1,214 @@
+
+                        // (lambda) * (lambda) = (A + x + x' + x'')
+                        let lambda_b = &lambda * &coeff_b;
+                        E::assert_eq(lambda_b * &lambda, xprime_lc);
+
+                        let yprime =
+                            witness!(|lambda, xprime, x_sum, y_sum| { -(y_sum + (lambda * (xprime - x_sum))) });
+
+                        E::assert_eq(lambda * (&x_sum - &xprime), &y_sum + &yprime);
+
+                        x_sum = xprime;
+                        y_sum = yprime;
+                    }
+                });
+
+                let edwards_x = &x_sum / y_sum; // 2 constraints
+                let edwards_y = (&x_sum - Field::one()) / (x_sum + Field::one()); // 2 constraints
+                Group::from_xy_coordinates(edwards_x, edwards_y) // 3 constraints
+            })
+            .fold(Group::zero(), |acc, group| acc + group)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use snarkvm_algorithms::{crh::BHPCRH as NativeBHP, CRH};
+    use snarkvm_circuits_environment::Circuit;
+    use snarkvm_curves::AffineCurve;
+    use snarkvm_utilities::{test_rng, UniformRand};
+
+    const ITERATIONS: usize = 10;
+    const MESSAGE: &str = "BHPCircuit0";
+    // const WINDOW_SIZE_MULTIPLIER: usize = 8;
+
+    type Projective = <<Circuit as Environment>::Affine as AffineCurve>::Projective;
+
+    fn check_hash<const NUM_WINDOWS: usize, const WINDOW_SIZE: usize>(
+        mode: Mode,
+        num_constants: usize,
+        num_public: usize,
+        num_private: usize,
+        num_constraints: usize,
+    ) {
+        // Initialize the BHP hash.
+        let native = NativeBHP::<Projective, NUM_WINDOWS, WINDOW_SIZE>::setup(MESSAGE);
+        let circuit = BHPCRH::<Circuit, NUM_WINDOWS, WINDOW_SIZE>::setup(MESSAGE);
+        // Determine the number of inputs.
+        let num_input_bits = NUM_WINDOWS * WINDOW_SIZE * BHP_CHUNK_SIZE;
+        // let num_input_bits = 128*8;
+
+        for i in 0..ITERATIONS {
+            // Sample a random input.
+            let input = (0..num_input_bits).map(|_| bool::rand(&mut test_rng())).collect::<Vec<bool>>();
+            // Compute the expected hash.
+            let expected = native.hash(&input).expect("Failed to hash native input");
+            // Prepare the circuit input.
+            let circuit_input: Vec<Boolean<_>> = Inject::new(mode, input);
+
+            Circuit::scope(format!("BHP {mode} {i}"), || {
+                // Perform the hash operation.
+                let candidate = circuit.hash(&circuit_input);
+                assert_eq!(expected, candidate.eject_value());
+
+                assert_scope!(num_constants, num_public, num_private, num_constraints);
+            });
+        }
+    }
+
+    // #[test]
+    // fn test_hash_constant() {
+    //     check_hash::<8, 32>(Mode::Constant);
+    // }
+
+    // #[test]
+    // fn test_hash_public() {
+    //     check_hash::<8, 32>(Mode::Public);
+    // }
```

Comment:
these should probably be uncommented or removed?

Reply:
Yeah, this PR is still a WIP